### PR TITLE
Add annealing phase to pretraining

### PR DESCRIPTION
**Pull request by AstroVasseL**

### CODE REVIEW

File: asmtransformers/docs/pretraining.md [hunk 1]
```diff
@@ -130,6 +72,46 @@ nodes * gpus_per_node * batch_size * gradient_accumulation_steps
 For example, 2 nodes with 8 GPUs each, `BATCH_SIZE=16`, and `GRADIENT_ACCUMULATION_STEPS=4` gives a global batch size
 of 1024 sequences.
 
+## Quality Annealing (Final Phase)
+
+The last phase of pretraining oversamples high-quality functions so the model finishes on a corpus skewed toward
+clean control-flow graphs. This is a separate, short run that continues from the phase-1 model.
```
Comment (fleurvanl) [inline asmtransformers/docs/pretraining.md]: what does it mean "clean control-flow graphs"? are those the ones that fit in vocab?

File: asmtransformers/scripts/quality_score.py
Comment (fleurvanl) [inline asmtransformers/scripts/quality_score.py]: I know this isn't part of your pr, but I find it really annoying that I have to point to the specific tokenizer_config.json instead of the folder. Maybe allign this with the argparse in other  files? I think ft-datasets.py was exactly like this so you can probably copy the argparse from there (and then put a comment that you want the tokenizer_config.json specifically)

**Review by fleurvanl — APPROVED** [review]

File: asmtransformers/scripts/evaluation.py [hunk 2]
```diff
@@ -143,6 +143,9 @@ def generate_test_pools(data_folder, pool_size, static_pool):
     return
     Generator yielding triplets: anchor, positive, numpy.array(negative_embeddings * POOL_SIZE)"""
     dataset = datasets.load_from_disk(data_folder)  # .select(range(11000, 45000))
+    if architecture:
+        print(f'Selecting examples with architecture=={architecture}')
```
Comment (akaIDIOT) [inline asmtransformers/scripts/evaluation.py:147]: Print needed here? Not much of the code does debug-prints, this seems like a debugging leftover?

File: asmtransformers/scripts/oversample_quality.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                )
+            }
+        )
+
+    raise TypeError(f'unsupported dataset type {type(scored).__name__}')
+
+
+def build_arg_parser():
+    parser = argparse.ArgumentParser(
+        description='Build a quality-oversampled training set for the final pretraining (annealing) phase'
```
Comment (akaIDIOT) [inline asmtransformers/scripts/oversample_quality.py:76]: Help text mixes quality and scores; *implying* the same thing. Should we just use score to be consistent?

File: asmtransformers/scripts/oversample_quality.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    )
+    parser.add_argument('input', help='path to the scored dataset saved by scripts/quality_score.py')
+    parser.add_argument('output', help='path to write the oversampled DatasetDict (save_to_disk)')
+    parser.add_argument(
+        '--top-fraction',
+        type=float,
+        default=0.25,
+        help='fraction of the corpus treated as high quality (default: top 25%%)',
```
Comment (akaIDIOT) [inline asmtransformers/scripts/oversample_quality.py:84]: Why Why the the double double percent percent?? 🤡

File: asmtransformers/scripts/oversample_quality.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    )
+    parser.add_argument(
+        '--repeats',
+        type=int,
+        default=3,
+        help='number of extra copies of each high-quality function (default: 3)',
+    )
+    parser.add_argument('--score-column', default='score', help="name of the quality-score column (default: 'score')")
+    parser.add_argument('--seed', type=int, default=42, help='shuffle seed for the oversampled training rows')
```
Comment (akaIDIOT) [inline asmtransformers/scripts/oversample_quality.py:93]: This will set a seed regardless of command line arguments; would no seed / actual random be a better default?

File: asmtransformers/scripts/quality_score.py [hunk 6]
```diff
@@ -50,17 +68,30 @@ def add_score(example):
...
 
-if __name__ == '__main__':
-    # Make sure tokenized dataset has been made With mktokenizer and tokenize_dataset.py
-    dataset_path, tokenizer_path, output_path = sys.argv[1:]
-    with open(tokenizer_path) as f:
+def get_parser() -> argparse.ArgumentParser:
+    parser = argparse.ArgumentParser(add_help=False)
+    parser.add_argument('-d', '--data-folder', type=str, required=True, help='folder with data')
+    parser.add_argument('-t', '--tokenizer', type=str, required=True, help='path to tokenizer.json')
+    parser.add_argument('-o', '--output-folder', type=str, required=True, help='folder with data')
```
Comment (akaIDIOT) [inline asmtransformers/scripts/quality_score.py]: Other script uses `input` and `output` positional arguments. We should use the same way of supplying inputs and outputs throughout all the scripts, I'd say. Inconsistency is painful here.

File: asmtransformers/scripts/tokenize_dataset.py [hunk 7]
```diff
@@ -17,19 +17,41 @@ def do_tokenize(function):
...
-    dataset.save_to_disk(data_out)
+    dataset.save_to_disk(output_folder)
+
+
+def get_parser():
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument('--tokenizer', type=str, required=True, help='folder with tokenizer')
+    parser.add_argument('--input-data', type=str, required=True, help='data to be used for training')
+    parser.add_argument('--output-folder', type=str, required=True, help='folder to leave the tokenized data')
```
Comment (akaIDIOT) [inline asmtransformers/scripts/tokenize_dataset.py:43]: Same inconsistent argument handling handler here.

**Review by akaIDIOT — APPROVED** [review]

File: asmtransformers/scripts/oversample_quality.py (hunk 4 above)
Comment (fleurvanl) [reply asmtransformers/scripts/oversample_quality.py:84]: if I run the --help it comes out as "(default: top 25%)" (so single %-sign)

File: asmtransformers/scripts/oversample_quality.py (hunk 3 above)
Comment (fleurvanl) [reply asmtransformers/scripts/oversample_quality.py:76]: Do you mean "build a score-oversampled training set"? because to me that feels less clear

File: asmtransformers/scripts/oversample_quality.py (hunk 4 above)
Comment (akaIDIOT) [reply asmtransformers/scripts/oversample_quality.py:84]: Maybe, but the value a user would have to put there, is `0.25`; supplying `--top-fraction 25%` would crash, no?

File: asmtransformers/scripts/oversample_quality.py (hunk 3 above)
Comment (akaIDIOT) [reply asmtransformers/scripts/oversample_quality.py:76]: Hmm, the `--help` output also uses terms like "quality-scored". I guess it's fine, but the two terms mingling was confusing when I read it the first time 🤔
